master/internal/api_workspace.go

```diff
@@ -551,6 +696,77 @@ func (a *apiServer) deleteWorkspace(
 	log.Debugf("workspace %d deleted successfully", workspaceID)
 }
 
+func (a *apiServer) ListWorkspaceNamespaceBindings(
+	ctx context.Context,
+	req *apiv1.ListWorkspaceNamespaceBindingsRequest,
+) (*apiv1.ListWorkspaceNamespaceBindingsResponse, error) {
+	curUser, _, err := grpcutil.GetUser(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	_, err = a.GetWorkspaceByID(ctx, req.Id, *curUser, false)
+	if err != nil {
+		return nil, err
+	}
+
+	var protoWorkspaceNamespaceBindings []*workspacev1.WorkspaceNamespace
+
+	// If a workspce does not have a workspace-namespace binding for a given cluster in the
+	// database, it is bound to the either the namespace specified for the corresponding resource
+	// manager in the master config or to the default Kubernetes namespace is the former field is
+	// empty.
+	for rmName := range a.m.allRms {
+		var wsns model.WorkspaceNamespace
+		protoBinding := &workspacev1.WorkspaceNamespace{
+			WorkspaceId:   req.Id,
+			NamespaceName: defaultKubernetesNamespace,
+			ClusterName:   rmName,
+		}
+
+		err := db.Bun().NewSelect().
+			Model(&model.WorkspaceNamespace{}).
+			Where("workspace_id = ?", req.Id).
+			Where("cluster_name LIKE ?", rmName).
+			Scan(ctx, &wsns)
+		if err != nil {
+			if err == sql.ErrNoRows {
+				protoWorkspaceNamespaceBindings = append(protoWorkspaceNamespaceBindings,
+					protoBinding)
+			} else {
+				return nil, fmt.Errorf("error getting workspace-namespace bindings: %w", err)
+			}
+		} else {
+			protoBinding = wsns.ToProto()
+			protoWorkspaceNamespaceBindings = append(protoWorkspaceNamespaceBindings, protoBinding)
+		}
+	}
+
+	// A workspace-namespace binding is considered "stale" if its cluster name is not used by any
+	// of the resource managers for the given determined deployment. List stale bindings as well,
+	// labeling them as such.
+	var workspaceNamespaceBindings []model.WorkspaceNamespace
+	err = db.Bun().NewSelect().
+		Model(&model.WorkspaceNamespace{}).
+		Where("workspace_id = ?", req.Id).
+		Scan(ctx, &workspaceNamespaceBindings)
+	if err != nil {
+		return nil, fmt.Errorf("error getting workspace-namespace bindings: %w", err)
+	}
+	for _, binding := range workspaceNamespaceBindings {
+		protoBinding := binding.ToProto()
+		if _, ok := a.m.allRms[binding.ClusterName]; !ok {
+			protoBinding.ClusterName += staleLabel
+			protoWorkspaceNamespaceBindings = append(protoWorkspaceNamespaceBindings,
+				protoBinding)
+		}
+	}
+
+	return &apiv1.ListWorkspaceNamespaceBindingsResponse{
```

Comment:
Do you think we could just return a string map for this function? Although this works fine, but in the web UI I will need to call this function and it would be better if it was a map instead of an object

Reply:
The response object is generated by protobuf, so we can't change it to a map. We do output the namespace bindings as an array though if that helps? It's a field of the response object

Reply:
Update post sync: Changed the type of the `NamespaceBindings` field from an array / repeated string to a `map<string, string>`!